### PR TITLE
adding import export session

### PR DESCRIPTION
**Pull request by StevenLAdler**
should it add users/items in the import when they dont already exist in DB?
should it auto clear on an import?
is 'unique' the correct field to match items?

tested and passing

### CODE REVIEW

**Comment by mandos21** [comment]
1. No, it should fail.
2. I think if we consider them as save/load, then yes it should clear beforehand
3. '_id' is the field you should match items on.

Ask chatgpt to write you some tests, usually that gets you 80-90% of the way there. That's what I've been doing and it still catches a bunch of issues.

File: app/views/session.py [hunk 1]
```diff
@@ -1,11 +1,54 @@
...
+        try:
+            user = User.objects.get(username=user_data['username'])
+            user.items = user_data['items']
+            user.save()
+            responses.add('200')
+        except DoesNotExist:
+            responses.add('400')
+
+    if '400' in responses:
+        return 'one or more invalid values', 400
```
Comment (mandos21) [inline app/views/session.py:48]: return jsonify({'message': 'One or more invalid fields'}), 400

File: app/views/session.py [hunk 2]
```diff
@@ -1,11 +1,54 @@
...
+    elif '200' in responses:
+        return 'import successful', 200
```
Comment (mandos21) [inline app/views/session.py:50]: I've just been returning 204s for successes that don't require any output, but I have no idea if that's a good practice.

**Review by mandos21 — CHANGES_REQUESTED** [review]

File: app/views/session.py [hunk 3]
```diff
@@ -1,11 +1,54 @@
...
+    session = {
+        "user": [user.to_mongo().to_dict() for user in User.objects()],
+        "item": [item.to_mongo().to_dict() for item in Item.objects()]
+    }
+    return json.dumps(session), 200
+
+
+@session_bp.route('/import', methods=['POST'])
+@token_required(dm_required=True)
+def import_session(**kwargs):
```
[the remark this answers is not in the document]
Comment (StevenLAdler) [reply app/views/session.py:24]: what if we just clear again on a partial failed import

File: app/views/session.py (hunk 2 above)
Comment (StevenLAdler) [reply app/views/session.py:50]: yeah thats right

File: app/views/session.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mandos21) [reply app/views/session.py:39]: works for me 👍

File: app/views/session.py [hunk 4]
```diff
@@ -24,14 +39,14 @@ def export_session(**kwargs):
...
         try:
-            item = Item.objects.get(unique=user_data['unique'])
+            item = Item.objects.get(_id=user_data['_id'])
             item.available = user_data['available']
             item.claimed = user_data['claimed']
             item.save()
-            responses.add('200')
+            responses.add('204')
         except DoesNotExist:
             responses.add('400')
```
Comment (mandos21) [inline app/views/session.py:52]: Might as well skip trying users if items failed

File: tests/views/test_session.py [hunk 5]
```diff
@@ -92,6 +93,83 @@ def test_clear_session_without_token(self):
         response = self.client.post('/session/clear')
 
         self.assertEqual(response.status_code, 403)  # Expecting Forbidden
+    
+    def test_export_session(client):
```
Comment (mandos21) [inline tests/views/test_session.py:97]: Should be self, not client

File: tests/views/test_session.py [hunk 6]
```diff
@@ -92,6 +93,83 @@ def test_clear_session_without_token(self):
...
+        # Add mock data
+        user = User(username="john_doe", items=["item1", "item2"]).save()
+        item = Item(name="Laptop", available=False, claimed=True).save()
+
+        # Call the export_session route
+        response = client.get('/export')
```
Comment (mandos21) [inline tests/views/test_session.py:103]: self.client.get

**Review by mandos21 — CHANGES_REQUESTED** [review]
The git workflow doesn't run pytests, I couldn't get that to work last I tried. You'll have to run pytest locally before pushing. There's a couple other small things in there to address as well.

**Comment by mandos21** [comment]
Also, I realize two of my comments weren't added to the review, but once I added them, I couldn't figure out how to do it without deleting them, so they're not in it technically?

File: app/views/session.py (hunk 4 above)
Comment (StevenLAdler) [reply app/views/session.py:52]: I'll just have it early return on failure

File: app/models/user.py [hunk 7]
```diff
@@ -21,3 +21,14 @@ def set_password(self, password):
 
     def check_password(self, password):
         return check_password_hash(self.password_hash, password)
+    
+    def get_item_names(self):
+        return [item.name for item in self.items]
+    
+    def to_dict(self):
```
Comment (mandos21) [inline app/models/user.py:28]: This function isn't necessary, the json provider in /app/utils/json_provider.py enhances the jsonify call to automatically deserialize items in a hash

File: tests/views/test_session.py [hunk 8]
```diff
@@ -1,30 +1,17 @@
...
-        cls.app = create_app()
-        cls.app.config['TESTING'] = True
-        cls.client = cls.app.test_client()
-
-    @classmethod
-    def tearDownClass(cls):
-        disconnect(alias='default')
+class SessionViewTestCase(ViewTestBase):
 
     def setUp(self):
```
Comment (mandos21) [inline tests/views/test_session.py:12]: You can replace everything from drop_all_collections() to self.token... with super().setUp()

**Review by mandos21 — CHANGES_REQUESTED** [review]
Looks good, just two small things

File: app/models/user.py (hunk 7 above)
Comment (mandos21) [reply app/models/user.py:28]: to_dict, that is. I realized that the preview in github makes it unclear which line I put the comment on.

**Pull request merged**